Add expected behavior to unified memory (#3617)
ROCm-Developer-Tools/HIP: docs/how-to/hip_runtime_api/memory_management/unified_memory.rst

.. meta::
  :description: This chapter describes Unified Memory and shows
                how to use it in AMD HIP.
  :keywords: AMD, ROCm, HIP, CUDA, unified memory, unified, memory

.. _unified_memory:

*******************************************************************************
Unified memory management
*******************************************************************************

In conventional architectures CPUs and attached devices have their own memory
space and dedicated physical memory backing it up, e.g. normal RAM for CPUs and

> VRAM on GPUs. This way each device can have physical memory optimized for its

Reviewer: ```suggestion
VRAM on GPUs. This way each device has physical memory optimized for its
```

Reviewer 2: "can" is actually correct here - there are "devices" that share their physical memory with the host (APUs in general, but explicitly the MI300A). But those support unified memory by default

> use case. Especially GPUs usually have specialized memory whose bandwidth is a

Reviewer: ```suggestion
use case. GPUs usually have specialized memory whose bandwidth is a
```

Reviewer 2: Added as suggested

magnitude higher than the RAM attached to CPUs.

While providing exceptional performance, this setup typically requires explicit
memory management, as memory needs to be allocated, copied and freed on the used
devices and on the host. Additionally, this makes using more than the physically
available memory on the devices complicated.

> Modern GPUs circumvent the problem of having to explicitly manage the memory,

Reviewer: Is this a feature of the "Modern GPUs" or of the runtime?

memory first and sent and mapped to host memory. This enables more memory to be
allocated and used for a GPU, than the GPU itself has physically available.

This level of unified memory support can be very beneficial for sparse accesses
to an array, that is not often used on the device.

Driver managed page migration
--------------------------------------------------------------------------------

If the hardware does not support on-demand page migration, then all the pages
accessed by a kernel have to be resident on the device, so they have to be
migrated before the kernel is running. Since the driver can not know beforehand,
what parts of an array are going to be accessed, all pages of all accessed
arrays have to be migrated. This can lead to significant delays on the first run
of a kernel, on top of possibly copying more memory than is actually accessed by
the kernel.

.. _unified memory system requirements:

System requirements
================================================================================

Unified memory is supported on Linux by all modern AMD GPUs from the Vega
series onward. Unified memory management can be achieved with managed memory

> allocation and, for the latest GPUs, with a system allocator.

Reviewer: ```suggestion
allocation using hipMallocManaged(), and for the latest GPUs with a system allocator, as shown in the following table.
```

Reviewer 2: Adapted it a bit

> The table below lists the supported allocators for the different architectures.

Reviewer: I think you can eliminate this paragraph after noting it in the prior paragraph.

Reviewer 2: Done

> The allocators are described in the next section.

Reviewer: ```suggestion

```

Reviewer 2: Done

`GPU memory <https://rocm.docs.amd.com/en/latest/conceptual/gpu-memory.html#xnack>`_.

.. _unified memory allocators:

Unified memory allocators
================================================================================

Support for the different unified memory allocators depends on the GPU
architecture and on the system. For more information, see :ref:`unified memory
system requirements` and :ref:`checking unified memory management support`.

- **HIP allocated managed memory and variables**
  
  :cpp:func:`hipMallocManaged()` is a dynamic memory allocator available on
  all GPUs with unified memory support. For more details, visit
  :ref:`unified_memory_reference`.

  The ``__managed__`` declaration specifier, which serves as its counterpart,
  can be utilized for static allocation.

- **System allocated unified memory**

  Starting with CDNA2, the ``new`` and `malloc()`` system allocators allow
  you to reserve unified memory. The system allocator is more versatile and
  offers an easy transition for code written for CPUs to HIP code as the
  same system allocation API is used.

To ensure the proper functioning of system allocated unified memory on supported

> graphics cards, it is essential to configure the environment variable

Reviewer: ```suggestion
accelerators and graphics cards, it is essential to configure the environment variable
```

Reviewer 2: There was no distinction between GPUs and accelerators in this whole chapter, and I won't start here, especially as the differences are miniscule at best

Reviewer: You differentiate between GPUs and graphic cards?

Reviewer 2: I don't think anyone would really see the terms "GPU" and "graphics card" as differentiation, they are synonyms. Changed it either way.

> ``XNACK=1`` and use a kernel that supports Heterogeneous Memory Management

Reviewer: Do we have a reference for Heterogeneous Memory Management?

Reviewer 2: It's a linux feature. I added a link

                // Copy input values to device.
                HIP_CHECK(hipMemcpy(d_a, &a, sizeof(*d_a), hipMemcpyHostToDevice));
                HIP_CHECK(hipMemcpy(d_b, &b, sizeof(*d_b), hipMemcpyHostToDevice));

                // Launch add() kernel on GPU.
                hipLaunchKernelGGL(add, dim3(1), dim3(1), 0, 0, d_a, d_b, d_c);

                // Copy the result back to the host.
                HIP_CHECK(hipMemcpy(&c, d_c, sizeof(*d_c), hipMemcpyDeviceToHost));

                // Cleanup allocated memory.
                HIP_CHECK(hipFree(d_a));
                HIP_CHECK(hipFree(d_b));
                HIP_CHECK(hipFree(d_c));

                // Prints the result.
                std::cout << a << " + " << b << " = " << c << std::endl;

                return 0;
            }

.. _using unified memory:

Using unified memory
================================================================================

Unified memory can simplify the complexities of memory management in GPU
computing, by not requiring explicit copies between the host and the devices. It
can be particularly useful in use cases with sparse memory accesses from both

> the CPU and the GPU, as not the whole memory region needs to be transferred to

Reviewer: ```suggestion
the CPU and the GPU, as the whole memory region does not need to be transferred to
```

Reviewer 2: There is a logical difference between "not the whole memory region needs to be transferred" and "the whole memory region does not need to be transferred"

Reviewer: What is the meaning of the sentence?

Reviewer 2: Changed the sentence to make it easier to understand.

PCIe bus.

In HIP, pinned memory allocations are coherent by default. Pinned memory is
host memory mapped into the address space of all GPUs, meaning that the pointer
can be used on both host and device. Addtionally, using pinned memory instead of
pageable memory on the host can improve bandwidth for transfers between the host
and the GPUs.

While unified memory can provide numerous benefits, it's important to be aware
of the potential performance overhead associated with unified memory. You must
thoroughly test and profile your code to ensure it's the most suitable choice
for your use case.

.. _unified memory runtime hints:

Performance optimizations for unified memory
================================================================================

There are several ways, in which the developer can guide the runtime to reduce
copies between devices, in order to improve performance.

Data prefetching
--------------------------------------------------------------------------------

Data prefetching is a technique used to improve the performance of your
application by moving data to the desired device before it's actually

> needed. ``hipCpuDeviceId`` is a special constant to specify the CPU as target.

Reviewer: This paragraph expresses that hipCpuDeviceId is a constant to specify the CPU as a target, but is it defined by the runtime? It does not seem to be established in the code example that follows?

Reviewer 2: I wouldn't say it is provided by the "runtime", it is a constant defined in the hip headers. I think this explanation here should also be good enough

        *a = 1;
        *b = 2;

        // Prefetch the data to the GPU device.
        HIP_CHECK(hipMemPrefetchAsync(a, sizeof(*a), deviceId, 0));
        HIP_CHECK(hipMemPrefetchAsync(b, sizeof(*b), deviceId, 0));
        HIP_CHECK(hipMemPrefetchAsync(c, sizeof(*c), deviceId, 0));

        // Launch add() kernel on GPU.
        hipLaunchKernelGGL(add, dim3(1), dim3(1), 0, 0, a, b, c);

        // Prefetch the result back to the CPU.
        HIP_CHECK(hipMemPrefetchAsync(c, sizeof(*c), hipCpuDeviceId, 0));

        // Wait for the prefetch operations to complete.
        HIP_CHECK(hipDeviceSynchronize());

        // Prints the result.
        std::cout << *a << " + " << *b << " = " << *c << std::endl;

        // Cleanup allocated memory.
        HIP_CHECK(hipFree(a));
        HIP_CHECK(hipFree(b));
        HIP_CHECK(hipFree(c));

        return 0;
    }

Memory advice
--------------------------------------------------------------------------------

Unified memory runtime hints can be set with :cpp:func:`hipMemAdvise()` to help
improve the performance of your code if you know the memory usage pattern. There
are several different types of hints as specified in the enum

> :cpp:enum:`hipMemoryAdvise`, e.g. whether a certain device mostly reads the

Reviewer: ```suggestion
:cpp:enum:`hipMemoryAdvise`, for example, whether a certain device mostly reads the
```

Reviewer 2: Done as suggested

        HIP_CHECK(hipMemAdvise(b, sizeof(*b), hipMemAdviseSetAccessedBy, deviceId));
        HIP_CHECK(hipMemAdvise(b, sizeof(*b), hipMemAdviseSetReadMostly, deviceId));

        // Set memory advice for c to be read, located on and accessed by the CPU.
        HIP_CHECK(hipMemAdvise(c, sizeof(*c), hipMemAdviseSetPreferredLocation, hipCpuDeviceId));
        HIP_CHECK(hipMemAdvise(c, sizeof(*c), hipMemAdviseSetAccessedBy, hipCpuDeviceId));
        HIP_CHECK(hipMemAdvise(c, sizeof(*c), hipMemAdviseSetReadMostly, hipCpuDeviceId));

        // Setup input values.
        *a = 1;
        *b = 2;

        // Launch add() kernel on GPU.
        hipLaunchKernelGGL(add, dim3(1), dim3(1), 0, 0, a, b, c);

        // Wait for GPU to finish before accessing on host.
        HIP_CHECK(hipDeviceSynchronize());

        // Prints the result.
        std::cout << *a << " + " << *b << " = " << *c << std::endl;

        // Cleanup allocated memory.
        HIP_CHECK(hipFree(a));
        HIP_CHECK(hipFree(b));
        HIP_CHECK(hipFree(c));

        return 0;
    }

Memory range attributes
--------------------------------------------------------------------------------

> :cpp:func:`hipMemRangeGetAttribute()` allows you to query attributes of a given

Reviewer: This is a statement of what you can do, but it might be a good idea to say why you might do it or when you might do it?

Reviewer 2: While I generally agree with that sentiment, this does not have a specific use case. It is really just forquerying the attributes of the memory ranges, that might have been set by the functions explained in the earlier sections. I can't think of an example that would be more than just showing how to call them.

        HIP_CHECK(hipDeviceSynchronize());

        // Query an attribute of the memory range.
        HIP_CHECK(hipMemRangeGetAttribute(&attributeValue,
                                attributeSize,
                                hipMemRangeAttributeReadMostly,
                                a,
                                sizeof(*a)));

        // Prints the result.
        std::cout << *a << " + " << *b << " = " << *c << std::endl;
        std::cout << "The array a is" << (attributeValue == 1 ? "" : " NOT") << " set to hipMemRangeAttributeReadMostly" << std::endl;

        // Cleanup allocated memory.
        HIP_CHECK(hipFree(a));
        HIP_CHECK(hipFree(b));
        HIP_CHECK(hipFree(c));

        return 0;
    }

Asynchronously attach memory to a stream
--------------------------------------------------------------------------------

The :cpp:func:`hipStreamAttachMemAsync()` function is able to asynchronously attach
memory to a stream, which can help concurrent execution when using streams.

Currently, this function is a no-operation (NOP) function on AMD GPUs. It simply
returns success after the runtime memory validation passed. This function is

> necessary on Microsoft Windows, and HMM is not supported on this operating

Reviewer: ```suggestion
required on Microsoft Windows, but HMM is not supported on this operating
```

It indicates that it is necessary on Microsoft Windows, but not needed on Linux? Therefore it doesn't do anything on Linux or isn't available on Linux? I understand HMM is not available on Windows, so using it on Windows results in nothing, but HMM is supported on Linux, so using it on Linux does nothing?

Reviewer 2: Good point, I didn't refactor this paragraph properly. I rewrote the description of what this function does to give a better understanding, and removed the paragraph about the support completely - HIP does not support unified memory on windows at all currently, and this function has not yet been implemented in Linux, so it is just for passing it along to CUDA based systems. That however is explained in the reference to this function, so it does not need to be repeated here.

series onward. Unified memory management can be achieved with managed memory
allocation and, for the latest GPUs, with a system allocator.

The table below lists the supported allocators for the different architectures.
The allocators are described in the next section.

.. list-table:: Supported Unified Memory Allocators

>     :widths: 40, 25, 25, 25

Reviewer 3: Now, you have three columns and not four
